src/test/librust/sapling_wallet_tests.cpp

```diff
@@ -0,0 +1,1133 @@
+                for (size_t j = 0; j < numBlocks; j++) {
+                    BOOST_CHECK((bool) saplingWitnesses[j]);
+                }
+                // Should equal final anchor because witness cache unaffected
+                BOOST_CHECK(saplingAnchors.back() == anchors);
+            }
+        }
+    }
+}
+
+BOOST_AUTO_TEST_CASE(ClearNoteWitnessCache) {
+    auto consensusParams = RegtestActivateSapling();
+
+    CWallet& wallet = *pwalletMain;
+    LOCK(wallet.cs_wallet);
+    setupWallet(wallet);
+
+    libzcash::SaplingExtendedSpendingKey sk = GetTestMasterSaplingSpendingKey();
+    BOOST_CHECK(wallet.AddSaplingZKey(sk));
+
+    CWalletTx wtx = GetValidSaplingReceive(Params().GetConsensus(),
+                                           wallet, sk, 10, true);
+    auto hash = wtx.GetHash();
+    auto saplingNotes = SetSaplingNoteData(wtx);
+
+    // Pretend we mined the tx by adding a fake witness
+    SaplingMerkleTree saplingTree;
+    wtx.mapSaplingNoteData[saplingNotes[0]].witnesses.push_front(saplingTree.witness());
+    wtx.mapSaplingNoteData[saplingNotes[0]].witnessHeight = 1;
+    wallet.GetSaplingScriptPubKeyMan()->nWitnessCacheSize = 1;
+
+    wallet.LoadToWallet(wtx);
+
+    // SetSaplingNoteData() only created a single Sapling output
+    // which is in the wallet, so we add a second SaplingOutPoint here to
+    // exercise the "note not in wallet" case.
+    saplingNotes.emplace_back(wtx.GetHash(), 1);
+    BOOST_CHECK_EQUAL(saplingNotes.size(), 2);
+
+    std::vector<Optional<SaplingWitness>> saplingWitnesses;
+
+    // Before clearing, we should have a witness for one note
+    GetWitnessesAndAnchors(wallet, saplingNotes, saplingWitnesses);
+    BOOST_CHECK((bool) saplingWitnesses[0]);
+    BOOST_CHECK(!(bool) saplingWitnesses[1]);
+    BOOST_CHECK_EQUAL(1, wallet.mapWallet[hash].mapSaplingNoteData[saplingNotes[0]].witnessHeight);
+    BOOST_CHECK_EQUAL(1, wallet.GetSaplingScriptPubKeyMan()->nWitnessCacheSize);
+
+    // After clearing, we should not have a witness for either note
+    wallet.GetSaplingScriptPubKeyMan()->ClearNoteWitnessCache();
+    auto anchors2 = GetWitnessesAndAnchors(wallet, saplingNotes, saplingWitnesses);
+    BOOST_CHECK(!(bool) saplingWitnesses[0]);
+    BOOST_CHECK(!(bool) saplingWitnesses[1]);
+    BOOST_CHECK_EQUAL(-1, wallet.mapWallet[hash].mapSaplingNoteData[saplingNotes[0]].witnessHeight);
+    BOOST_CHECK_EQUAL(0, wallet.GetSaplingScriptPubKeyMan()->nWitnessCacheSize);
+}
+
+BOOST_AUTO_TEST_CASE(UpdatedSaplingNoteData) {
+    auto consensusParams = RegtestActivateSapling();
+
+    CWallet& wallet = *pwalletMain;
+    LOCK(wallet.cs_wallet);
+    setupWallet(wallet);
+
+    auto m = GetTestMasterSaplingSpendingKey();
+
+    // Generate dummy Sapling address
+    auto sk = m.Derive(0);
+    auto expsk = sk.expsk;
+    auto extfvk = sk.ToXFVK();
+    auto pa = sk.DefaultAddress();
+
+    // Generate dummy recipient Sapling address
+    auto sk2 = m.Derive(1);
+    auto expsk2 = sk2.expsk;
```

Comment:
(9cabc2ed2399221b23491c536254488dda44e60e)
It seems that  this variable is never used.